lib/chef/provider/package/dnf.rb

```diff
@@ -0,0 +1,298 @@
+        extend Chef::Mixin::Which
+        include Chef::Mixin::GetSourceFromPackage
+
+        allow_nils
+
+        class Version
+          attr_accessor :name
+          attr_accessor :version
+          attr_accessor :arch
+
+          def initialize(name, version, arch)
+            @name = name
+            @version = ( version == "nil" ) ? nil : version
+            @arch = ( arch == "nil" ) ? nil : arch
+          end
+
+          def to_s
+            "#{name}-#{version}.#{arch}"
+          end
+
+          def version_with_arch
+            "#{version}.#{arch}" unless version.nil?
+          end
+
+          def matches_name_and_arch?(other)
+            other.version == version && other.arch == arch
+          end
+
+          def ==(other)
+            name == other.name && version == other.version && arch == other.arch
+          end
+
+          alias_method :eql?, :==
+        end
+
+        attr_accessor :python_helper
+
+        class PythonHelper
+          include Singleton
+          extend Chef::Mixin::Which
+
+          attr_accessor :stdin
+          attr_accessor :stdout
+          attr_accessor :stderr
+          attr_accessor :wait_thr
+
+          DNF_HELPER = ::File.expand_path(::File.join(::File.dirname(__FILE__), "dnf_helper.py")).freeze
+          DNF_COMMAND = "#{which("python3")} #{DNF_HELPER}"
+
+          def start
+            ENV["PYTHONUNBUFFERED"] = "1"
+            @stdin, @stdout, @stderr, @wait_thr = Open3.popen3(DNF_COMMAND)
+          end
+
+          def reap
+            unless wait_thr.nil?
+              Process.kill("KILL", wait_thr.pid) rescue nil
+              stdin.close unless stdin.nil?
+              stdout.close unless stdout.nil?
+              stderr.close unless stderr.nil?
+              wait_thr.value
+            end
+          end
+
+          def check
+            start if stdin.nil?
+          end
+
+          # i couldn't figure out how to decompose an evr on the python side, it seems reasonably
+          # painless to do it in ruby.
+          def add_version(hash, version)
+            epoch = nil
+            if version =~ /(\S+):(\S+)/
+              epoch, version = $1, $2
+            end
+            if version =~ /(\S+)-(\S+)/
+              version, release = $1, $2
+            end
+            hash["epoch"] = epoch unless epoch.nil?
+            hash["release"] = release unless release.nil?
+            hash["version"] = version
+          end
+
+          # @returns Array<Version>
+          def query(action, provides, version = nil, arch = nil)
+            with_helper do
+              hash = { "action" => action }
+              hash["provides"] = provides
+              add_version(hash, version) unless version.nil?
+              hash["arch" ] = arch unless arch.nil?
+              json = FFI_Yajl::Encoder.encode(hash)
+              puts json
+              stdin.syswrite json + "\n"
+              output = stdout.sysread(4096)
```

Comment:
Yeah, I'm not too worried about overflowing the 4k buffer because we only ever send one request at a time and the name+version+arch on the new resource is unlikely to ever get that large.  On the way back we also only ever receive one package to install (all the filtering down to the one package nevra happens python-side -- even though the ruby code would handle multiple versions, it never actually receives them).

And I think I have the same feelings about this as the file descriptor which is that it makes it more difficult to play with the helper on the command line.